### PR TITLE
simplify credentials module

### LINKED ISSUE #956: Unable to use AWS ECS task role credentials
IAM provider source code: https://github.com/minio/minio-py/blob/1ee647eaf9507e04b21404a251bc6953b9861c13/minio/credentials/providers.py#L187

The issue is that `creds_path = "/latest/meta-data/iam/security-credentials"` is hardcoded.

On an EC2 instance, you can run `curl http://169.254.169.2data/iam/security-credentials` to get the iam role, then `curl http://169.254.169.2data/iam/security-credentials/ROLENAME` to get the credentials.

For an ECS task running with a task iam role on an ec2 instance, you can get credentials if you run `curl 169.254.170.2$AWS_CONTAINER_CREDENTIALS_RELATIVE_URI`.  The value for `AWS_CONTAINER_CREDENTIALS_RELATIVE_URI` is typically like: `/v2/credentials/54da6c7a-84ff-41dc-a160-21b8a8bc7fb0`. The provider lets you set `endpoint`, but since `creds_path` is hardcoded, I cannot get the credentials.

Also, the logic for "Get credentials of first role." is not required in the ECS task method.

I suspect a fix would involve either making `creds_path` also configurable (and also the logic to get the first role, since it directly gives the credentials), or checking if `AWS_CONTAINER_CREDENTIALS_RELATIVE_URI` is present in the environment variables and changing behavior accordingly.

Example results from `curl 169.254.170.2$AWS_CONTAINER_CREDENTIALS_RELATIVE_URI`:

```
{"RoleArn":"arn:aws:iam::ACCOUNTID:role/ROLENAME","AccessKeyId":"ASIAXXXX","SecretAccessKey":"XXXXXX","Token":"TOKENTEXT","Expiration":"2020-08-13T06:11:55Z"}
```

### PR DESCRIPTION
**Pull request by balamurugana**
* Added LDAP, ClientGrants and WebIdentity providers.
* Refactor IAM provider and renamed to IamAwsProvider.
* Refactor AssumeRole provider.
* Renamed FileAWSCredentials provider to AWSConfigProvider.
* Renamed Chain provider to ChainedProvider.
* Renamed EnvAWS provider to EnvAWSProvider.
* Renamed EnvMinio provider to EnvMinioProvider.
* Renamed FileMinioClient provider to MinioClientConfigProvider.
* Renamed Static provider to StaticProvider.

Fixes #956 #957

### CODE REVIEW

File: minio/credentials/providers.py [hunk 1]
```diff
@@ -17,70 +17,181 @@
...
-        self._expiry = None
+    def __init__(
+            self, sts_endpoint, access_key, secret_key, duration_seconds=0,
+            policy=None, region=None, role_arn=None, role_session_name=None,
+            external_id=None, http_client=None,
+    ):
+        self._sts_endpoint = sts_endpoint
+        self._access_key = access_key
+        self._secret_key = secret_key
+        self._region = region
```
Comment (Praveenrajmani) [inline minio/credentials/providers.py:98]: shouldn't we fallback to "us-east-1" as default if `region` is not specified?

**Review by Praveenrajmani — APPROVED** [review]
Just few questions/doubts. Looks good to me otherwise

File: minio/api.py [hunk 2]
```diff
@@ -154,22 +153,13 @@ def __init__(self, endpoint, access_key=None,
...
-            provider=Chain(
-                providers=[
-                    Static(access_key, secret_key, session_token),
-                    EnvAWS(),
-                    EnvMinio(),
-                ]
-            )
-        )
+        if access_key:
+            credentials = StaticProvider(access_key, secret_key, session_token)
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply minio/api.py:161]: Yes. This PR breaks backward compatibility to some extends. Previously getting credentials from env is wrong i.e. there is no way to create `Minio` with anonymous access unless `Chain` provider returns `None` credentials. This looks odd and no control to users.

As credentials is concerned, the user should have more control over by passing necessary credential provider. Additionally when access/secret keys are provided along with credential provider, its better to ignore credential provider and use access/secret keys (as static credentials). This is what I changed.

Note that Chain provider are meant to give valid access/secret keys, they are not credential fail-over mechanism i.e. if one access/secret keys are invalid, then fetch another access/secret from next provider.

File: minio/credentials/providers.py (hunk 1 above)
Comment (balamurugana) [reply minio/credentials/providers.py:98]: I see `region` is optional in `minio-go`; however for signature v4 region must be needed. I will check why `region` is optional in `minio-go` and fix accordingly here.

Comment (balamurugana) [reply minio/credentials/providers.py:98]: I got clarified on this. We would need to use empty string as region if not provided. Fixing accordingly.

File: minio/api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Praveenrajmani) [reply minio/api.py:161]: understood. Thanks!

**Review by Praveenrajmani — APPROVED** [review]
LGTM

**Review by nitisht — APPROVED** [review]

**Pull request merged**